crates/data_components/src/s3_vectors/partition.rs

```diff
@@ -0,0 +1,231 @@
+        let parts: Vec<&str> = index_name.split('-').collect();
+        let num_parts = parts.len();
+        ensure!(num_parts == 4, IncorrectNumPartsInNameSnafu { num_parts });
+        Ok(Self {
+            index_name: parts[0].to_string(),
+            column_name: parts[1].to_string(),
+            partition_value_hash: parts[2].to_string(),
+            partition_by_hash: parts[3].to_string(),
+        })
+    }
+
+    /// Determines if the partitions come from the same dataset
+    #[must_use]
+    pub fn belongs_with(&self, other: &Self) -> BelongsWith {
+        if self.index_name != other.index_name || self.column_name != other.column_name {
+            return BelongsWith::DifferentDataset;
+        }
+        if self.partition_by_hash != other.partition_by_hash {
+            return BelongsWith::DifferentParitionByExpressions;
+        }
+        BelongsWith::SameDataset
+    }
+}
+
+fn sanitize(s: &str) -> String {
+    s.replace('_', "-")
+}
+
+fn truncate(s: &str, len: usize) -> String {
+    s.chars().take(len).collect()
+}
+
+fn hash_to_hex(input: &str) -> String {
+    let hash = XxHash64::oneshot(HASH_SEED, input.as_bytes());
+    format!("{hash:x}")
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use datafusion::prelude::col;
+    use datafusion::scalar::ScalarValue;
+
+    /// See [CreateIndex](https://docs.aws.amazon.com/AmazonS3/latest/API/API_S3VectorBuckets_CreateIndex.html#API_S3VectorBuckets_CreateIndex_RequestSyntax)
+    const S3_VECTOR_INDEX_NAME_MAX_LENGTH: usize = 63;
+
+    #[test]
+    fn index_name_length_restricted() {
+        let index_name = "a".repeat(INDEX_NAME_MAX_LENGTH + 1);
+        let column_name = "col1";
+        let partition_value = ScalarValue::from("val");
+        let partition_by = vec![col("col1")];
+
+        let result =
+            PartitionedIndexName::new(&index_name, column_name, &partition_value, &partition_by)
+                .expect("result");
+        let result = result.to_index_name();
+
+        assert!(result.len() <= S3_VECTOR_INDEX_NAME_MAX_LENGTH);
+        assert_eq!(
+            result.split('-').next().expect("next").len(),
+            INDEX_NAME_MAX_LENGTH
+        );
+    }
+
+    #[test]
+    fn new_index_partition_name() {
+        let index_name = "test_index";
+        let column_name = "test_col";
+        let partition_value = ScalarValue::from("value");
+        let partition_by = vec![col("col1")];
+
+        let result =
+            PartitionedIndexName::new(index_name, column_name, &partition_value, &partition_by)
+                .expect("new");
+
+        assert_eq!(result.index_name, "test-index");
+        assert_eq!(result.column_name, "test-col");
+        assert_eq!(
+            result.partition_value_hash.len(),
+            PARTITION_VALUE_MAX_LENGTH
+        );
+        assert_eq!(result.partition_by_hash.len(), PARTITION_BY_MAX_LENGTH);
+    }
+
+    #[test]
+    fn from_index_name_valid() {
```

Comment:
Let's add a test where the index name and column name has a `-`, i.e. `test-index-test-col-12345-abcde`